### PR TITLE
Properly ban the negation of unsigned integers in type-checking.

### PR DESCRIPTION
**Pull request by eddyb**
Lint-time banning of unsigned negation appears to be vestigial from a time it was feature-gated.
But now it always errors and we do have the ability to deref the checking of e.g. `-0`, through the trait obligation fulfillment context, which will only succeed/error when the `0` gets inferred to a specific type.

The two removed tests are the main reason for finally cleaning this up, they need changing all the time when refactoring the HIR-based `rustc_const_eval` and/or `rustc_passes::consts`, as warnings pile up.

### CODE REVIEW

**Comment by nagisa** [comment]
How complicated is it to retain the `HELP use !` message? People of C background do `-1` for what really is a `!0` all the time and will have a hard time figuring out the binary not by themselves 100% of the time.

File: src/test/compile-fail/feature-gate-negate-unsigned0.rs [hunk 1]
```diff
@@ -1,34 +0,0 @@
...
-
-    let _d = -1u8;
-    //~^ ERROR E0080
-    //~| unary negation of unsigned integer
-
-    for _ in -10..10u8 {}
-    //~^ ERROR E0080
-    //~| unary negation of unsigned integer
-
-    -S; // should not trigger the gate; issue 26840
```
Comment (nagisa) [inline src/test/compile-fail/feature-gate-negate-unsigned0.rs]: This is a regression test for #26840. Perhaps should be retained in some way?

Comment (nagisa) [inline src/test/compile-fail/feature-gate-negate-unsigned0.rs]: Oh wait, what? Why is there a duplicate file?

Comment (eddyb) [inline src/test/compile-fail/feature-gate-negate-unsigned0.rs]: This is the problematic split. It has to do with how there's typeck, followed by lints and then followed by trans, and lints themselves can have these errors and optionally const eval warnings.
I think @oli-obk knows exactly why this was needed in the first place, but now there is no point.

**Comment by eddyb** [comment]
@nagisa You'd want to do that on `uN: Neg` not implemented, which is something we should do in general (have a way to customize error messages for certain combination of unimplemented types).

**Comment by eddyb** [comment]
Travis error should be fixed by rust-lang/libc#484, in theory, waiting for a merge of that.

**Comment by pnkfelix** [comment]
@eddyb wrote:

> You'd want to do that on uN: Neg not implemented, which is something we should do in general (have a way to customize error messages for certain combination of unimplemented types).

Hmm, was PR #33401 expressive enough to cover this scenario?

**Comment by pnkfelix** [comment]
Okay, I was worried at first when I saw the two tests being removed, but now I understand that there remains the variants that do not have the `0` suffix.

But ...

@eddyb: shouldn't `const-eval-overflow.rs` be updated with cases for all the `-uN::MIN` that are being removed here? i.e. undo the effect of bac3eec67c367718c05585b72974c58a66f6abd0 for this file?

**Comment by pnkfelix** [comment]
r=me after travis passes.

**Pull request merged**